### PR TITLE
Implementation of JsBox

### PR DESCRIPTION
**Pull request by kjvalencik**
Implements [`JsBox` RFC](https://github.com/neon-bindings/rfcs/pull/33).

### CODE REVIEW

**Comment by jrose-signal** [comment]
What's the status of this PR? Is there anything blocking it at this point?

**Comment by kjvalencik** [comment]
@jrose-signal It still needs a review from @dherman and/or @goto-bus-stop.

File: src/types/boxed.rs [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+/// The type `JsBox<T>` provides shared ownership of a value of type `T`,
+/// allocated in the heap. The data is owned by the JavaScript engine and the
+/// lifetime is managed by the JavaScript garbage collector.
+///
+/// Shared references in Rust disallow mutation by default, and `JsBox` is no
+/// exception: you cannot generally obtain a mutable reference to something
+/// inside a `JsBox`. If you need to mutate through a `JsBox`, use
+/// [`Cell`](https://doc.rust-lang.org/std/cell/struct.Cell.html),
+/// [`RefCell`](https://doc.rust-lang.org/stable/std/cell/struct.RefCell.html),
+/// or one of the other types that provide interior mutability.
```
Comment (dherman) [inline src/types/boxed.rs:26]: Let's link the phrase "interior mutability" to the [book chapter on interior mutability](https://doc.rust-lang.org/book/ch15-05-interior-mutability.html).

File: src/types/boxed.rs [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+/// # use neon::prelude::*;
+/// # use std::cell::RefCell;
+///
+/// type BoxedPerson = JsBox<RefCell<Person>>;
+///
+/// struct Person {
+///      name: String,
+/// }
+///
+/// impl Finalize for Person {}
```
Comment (dherman) [inline src/types/boxed.rs:79]: I think we haven't documented the `Finalize` trait and how it's connected to `JsBox` anywhere in this PR?

File: src/types/boxed.rs [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+
+        v.and_then(|v| v.downcast_ref())
+            .map(|internal| Self {
+                local,
+                internal,
+            })
+    }
+}
+
+impl<T: Finalize + Send + 'static> JsBox<T> {
```
Comment (dherman) [inline src/types/boxed.rs:192]: Can we explain in the rustdocs why each trait and lifetime bound is here?

File: src/types/boxed.rs [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+    /// Constructs a new `JsBox` containing `value`.
+    pub fn new<'a, C>(cx: &mut C, value: T) -> Handle<'a, JsBox<T>>
+    where
+        C: Context<'a>,
+        T: Send + 'static,
+    {
+        fn finalizer<U: Finalize + 'static>(env: raw::Env, data: BoxAny) {
```
Comment (dherman) [inline src/types/boxed.rs:199]: Can we document in a comment what this function is for? I think the idea is that it runs the finalization of the Rust data whenever the GC decides to collect the `JsBox` wrapper object, yes?

**Review by dherman — CHANGES_REQUESTED** [review]
This looks like a pretty clean implementation! There's just some parts that are a little dense and hard to follow, so most of my review comments are about documenting and clarifying the code.

File: src/types/boxed.rs [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+///     let greeting = person.greet();
+/// 
+///     Ok(cx.string(greeting))
+/// }
+pub struct JsBox<T: Send + 'static> {
+    local: raw::Local,
+    // `JsBox` cannot verify the lifetime. Store a raw pointer to force uses
+    // to be marked unsafe. In practice, it can be treated as `'static` but
+    // should only be exposed as part of a `Handle` tied to a `Context` lifetime.
+    internal: *const T,
```
[the remark this answers is not in the document]
Comment (kjvalencik) [reply src/types/boxed.rs:124]: I can add a comment. It's a pointer to the user's data.

File: src/types/boxed.rs [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+use neon_runtime::raw;
+use neon_runtime::external;
+
+use crate::context::{Context, FinalizeContext};
+use crate::context::internal::Env;
+use crate::handle::{Managed, Handle};
+use crate::types::internal::ValueInternal;
+use crate::types::Value;
+
+type BoxAny = Box<dyn Any + Send + 'static>;
```
[the remark this answers is not in the document]
Comment (kjvalencik) [reply src/types/boxed.rs:13]: I don't think it makes sense to add parenthesis. These are trait bounds, so they all apply equally. `dyn` is the "optional" specifier that was added in Rust 2018.

In my opinion, adding parenthesis would be less idiomatic.

File: src/types/boxed.rs [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+    }
+}
+
+impl<T: Send + 'static> ValueInternal for JsBox<T> {
+    fn name() -> String {
+        any::type_name::<Self>().to_string()
+    }
+
+    fn is_typeof<Other: Value>(env: Env, other: Other) -> bool {
+        let v = unsafe {
```
[the remark this answers is not in the document]
Comment (kjvalencik) [reply src/types/boxed.rs:169]: I think a helper method would make this less clear. `.map(|v| &*v)` is simply converting a raw pointer `*` to `&`.

File: src/types/boxed.rs [hunk 8]
```diff
@@ -0,0 +1,343 @@
...
+            external::deref::<BoxAny>(env.to_raw(), other.to_raw())
+                .map(|v| &*v)
+        };
+
+        v.map(|v| v.is::<T>()).unwrap_or(false)
+    }
+
+    fn downcast<Other: Value>(env: Env, other: Other) -> Option<Self> {
+        let local = other.to_raw();
+        let v = unsafe {
```
[the remark this answers is not in the document]
Comment (kjvalencik) [reply src/types/boxed.rs:179]: I can see your argument that this code is repeated several times, but it's also _dereferencing a raw pointer_ which is `unsafe` and I am concerned about obscuring this safety. If I move it to a helper, I would document the safety in each location; it wouldn't be more terse.

File: src/types/boxed.rs [hunk 9]
```diff
@@ -0,0 +1,343 @@
...
+
+            FinalizeContext::with(
+                env,
+                move |mut cx| data.finalize(&mut cx),
+            );
+        }
+
+        let v = Box::new(value) as BoxAny;
+        // Since this value was just constructed, we know it is `T`
+        let internal = &*v as *const dyn Any as *const T;
```
[the remark this answers is not in the document]
Comment (kjvalencik) [reply src/types/boxed.rs:211]: `internal` is the user's data. Maybe it would be helpful to rename the field? This code is taking `Box<Any>` and getting a `*const T` from it.

File: src/types/boxed.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (dherman) [reply src/types/boxed.rs:124]: I think the things that confused me were:

1. It's the only `JsFoo` type that isn't just a zero-cost wrapper of `raw::Local`.
2. I didn't understand that this is just a cached copy of the extracted and downcast underlying pointer, i.e. just an optimization. It would be helpful to call that out.
3. The fact that it's wider than other `JsFoo` types means it can't be used as a `This` type.

I think if we just document this with internal comments that would be helpful.

File: src/types/boxed.rs [hunk 10]
```diff
@@ -0,0 +1,343 @@
...
+
+impl<T: Send + 'static> Value for JsBox<T> { }
+
+impl<T: Send + 'static> Managed for JsBox<T> {
+    fn to_raw(self) -> raw::Local {
+        self.local
+    }
+
+    fn from_raw(env: Env, h: raw::Local) -> Self {
+        let v = unsafe {
```
[the remark this answers is not in the document]
Comment (dherman) [reply src/types/boxed.rs:147]: Maybe this would be a little easier to follow, and with slightly more specific error messages, if we unwrap the optional results at each step? Something like:

```rust
fn from_raw(env: Env, h: raw::Local) -> Self {
    let p: *const T = external::deref::<BoxAny>(env.to_raw(), h)
        .expect("JsBox failed to unwrap external");
    let r: &T = &*p;
    let unwrapped = r.downcast_ref()
        .expect("JsBox failed to downcast external");

    Self {
        local: h,
        unwrapped
    }
}
```

File: src/types/boxed.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (dherman) [reply src/types/boxed.rs:179]: What if we make the helper an `unsafe fn`?

File: src/types/boxed.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (kjvalencik) [reply src/types/boxed.rs:124]: Per slack discussion, I will be removing the cached pointer optimization. It will be replaced by a `PhantomData`. Any preference on name? I usually go with something generic like `_phantom` or `_marker`.

File: src/types/boxed.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (kjvalencik) [reply src/types/boxed.rs:179]: This will be clearer and less necessary once the cached pointer is removed.

File: src/types/boxed.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (kjvalencik) [reply src/types/boxed.rs:124]: After further thought, the cached pointer is necessary to facilitate `Deref` without needing an `Env`. I will add additional documentation.

File: src/types/boxed.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (dherman) [reply src/types/boxed.rs:13]: So that means it's `dyn (Any + Send + 'static)`? That's fine if it's not idiomatic to parenthesize. I'd just like to understand.

[the remark this answers is not in the document]
Comment (kjvalencik) [reply src/types/boxed.rs:13]: Parenthesis can't change the way that these are applied. It's always all 3 on this simple type. I think with parenthesis it would read `(dyn (Any + Send)) + 'static`.

The only time you typically see parenthesis is with closures to distinguish whether a lifetime is associated with the return type or the closure.

**Review by dherman — APPROVED** [review]
Thanks for the comments, it's much easier to follow now! :shipit:
